pkg/resources/process/wait.go

```diff
@@ -69,47 +72,101 @@ func (w *procWaiter) wait(ch chan<- *drivers.ExitResult) {
 	}
 }
 
+const (
+	// exitStatusFile is written into the task directory upon the exit of the
+	// parent task process. It contains the exit status in case the value needs
+	// to be retrieved by the plugin (i.e. if the task process was orphaned).
+	exitStatusFile = ".exit_status.txt"
+)
+
 // WaitPID is able to wait on a given specific PID. We must lookup the
 // process and also send a signal(0) to make sure it is actually still alive
 // before waiting on it.
-func WaitPID(pid int) Waiter {
+func WaitPID(pid int, taskdir string) Waiter {
 	return &pidWaiter{
-		pid: pid,
-		ch:  make(chan *drivers.ExitResult),
+		pid:     pid,
+		taskdir: taskdir,
+		ch:      make(chan *drivers.ExitResult),
 	}
 }
 
 type pidWaiter struct {
-	pid int
-	ch  chan *drivers.ExitResult
+	pid     int
+	taskdir string
+	ch      chan *drivers.ExitResult
 }
 
 func (w *pidWaiter) Wait() WaitCh {
 	go w.wait(w.ch)
 	return w.ch
 }
 
-func (w *pidWaiter) wait(ch chan<- *drivers.ExitResult) {
-	proc, _ := os.FindProcess(w.pid) // never errors on unix
+func unrecoverable(msg string, err error) *drivers.ExitResult {
+	return &drivers.ExitResult{
+		ExitCode: -1,
+		Err:      fmt.Errorf("%s: %w", msg, err),
+	}
+}
 
+func (w *pidWaiter) wait(ch chan<- *drivers.ExitResult) {
 	// send a 0 signal to detect if the process is still alive
+	proc, _ := os.FindProcess(w.pid) // never errors on unix
 	if err := proc.Signal(syscall.Signal(0)); err != nil {
-		// TODO(shoenig): there is still a way to get the exit code and signal
-		// of a recently deceased process by reading /proc; we should look into
-		// that after Nomad 1.8 Beta since Client restarts are a thing that we
-		// support.
-		//
-		// See what the pledge driver is doing for inspiration in
-		// https://github.com/shoenig/nomad-pledge-driver/blob/v0.3.0/pkg/resources/process/wait.go#L82-L105
-		// and make sure it works correctly.
-		ch <- &drivers.ExitResult{
-			ExitCode: -1,
-			Err:      fmt.Errorf("task is gone: %w", err),
-		}
+		ch <- unrecoverable("task process handle is gone", err)
+		return
+	}
+
+	pidfd, err := openProcessFD(w.pid)
+	if err != nil {
+		ch <- unrecoverable("task process file descriptor cannot be opened", err)
+		return
+	}
```

Comment:
If the task exits quickly (ex the shim process fails during startup), it'll have written any to the named pipe logs but we won't have the exit code? I feel like what I'm asking is supposed to be covered given your description:

> wait.go now splits logic between waiting on a child and waiting on an orphan.

But I'm not sure I understand where that split is happening here.